File: testdata/precision-suite-lua/tp_pathtraversal.lua
-- Path traversal (CWE-22): an untrusted value is used as the path argument of
-- io.open, so `../` sequences escape the intended directory. A correct scanner
-- fires TAINT-004.

-- A CLI argument opened directly as a file path.
local name = arg[1]
local f = io.open("/var/app/data/" .. name) -- nox-expect: TAINT-004

-- An OpenResty request argument opened as a path.
local rel = ngx.req.get_uri_args().file
local fh = io.open(rel) -- nox-expect: TAINT-004

File: testdata/precision-suite-scala/tp_pathtraversal.scala
// Path traversal: a request parameter is concatenated into a filesystem path
// opened via scala.io.Source.fromFile — CWE-22. A correct scanner fires
// TAINT-004. The untrusted name is neither canonicalized nor stripped of `..`
// components, so it can escape the intended directory.
package controllers

import scala.io.Source

class DownloadController {
  // read opens a file whose name comes directly from untrusted input — the bug.
  def read(request: Request): String = {
    val name = request.getQueryString("file")
    val path = "/srv/reports/" + name
    Source.fromFile(path).mkString // nox-expect: TAINT-004
  }
}
